docs/eBpfExtensions.md

```diff
@@ -1,42 +1,207 @@
+    uint64_t data_meta; ///< Packet metadata.
+
+    /* size: 12, cachelines: 1, members: 3 */
+    /* last cacheline: 12 bytes */
+} xdp_md_t;
+```
+The corresponding context descriptor looks like:
 ```
-ebpf_error_code_t (*invoke_hook)(void* bind_context, void* context, uint32_t* result);
+const ebpf_context_descriptor_t g_xdp_context_descriptor = {sizeof(xdp_md_t),
+                                                            EBPF_OFFSET_OF(xdp_md_t, data),
+                                                            EBPF_OFFSET_OF(xdp_md_t, data_end),
+                                                            EBPF_OFFSET_OF(xdp_md_t, data_meta)};
 ```
+If any of the data or metadata pointer fields are not present on the context structure, the offset value is set to -1 in the context descriptor.
+
+#### `ebpf_helper_function_prototype_t` Struct
+This structure is used to describe the prototypes of the various helper functions implemented by the extension.
+* `helper_id`: Integer signifying the helper function Id. (See section 3.3).
+Helper function IDs for different program types may not be unique.
+* `name`: Helper function name.
+* `return_type`: Set the appropriate value for `ebpf_return_type_t` enum that represents the return type of the helper function.
+* `arguments`: Array of (at most) five helper function arguments of type `ebpf_argument_type_t`.
+
+#### `ebpf_argument_type_t` Enum
+This enum describes the various argument types that can be passed to an eBPF helper function. This is defined in [Prevail Verifier](https://github.com/vbpf/ebpf-verifier) project.
+```
+typedef enum _ebpf_argument_type {
+    EBPF_ARGUMENT_TYPE_DONTCARE = 0,
+    EBPF_ARGUMENT_TYPE_ANYTHING,
+    EBPF_ARGUMENT_TYPE_CONST_SIZE,
+    EBPF_ARGUMENT_TYPE_CONST_SIZE_OR_ZERO,
+    EBPF_ARGUMENT_TYPE_PTR_TO_CTX,
+    EBPF_ARGUMENT_TYPE_PTR_TO_MAP,
+    EBPF_ARGUMENT_TYPE_PTR_TO_MAP_OF_PROGRAMS,
+    EBPF_ARGUMENT_TYPE_PTR_TO_MAP_KEY,
+    EBPF_ARGUMENT_TYPE_PTR_TO_MAP_VALUE,
+    EBPF_ARGUMENT_TYPE_PTR_TO_MEM,
+    EBPF_ARGUMENT_TYPE_PTR_TO_MEM_OR_NULL,
+    EBPF_ARGUMENT_TYPE_PTR_TO_UNINIT_MEM,
+} ebpf_argument_type_t;
+```
+
+#### `ebpf_return_type` Enum
+This enum describes the various return types from an eBPF helper function. This is defined in [Prevail Verifier](https://github.com/vbpf/ebpf-verifier) project.
+```
+typedef enum _ebpf_return_type {
+    EBPF_RETURN_TYPE_INTEGER = 0,
+    EBPF_RETURN_TYPE_PTR_TO_MAP_VALUE_OR_NULL,
+    EBPF_RETURN_TYPE_INTEGER_OR_NO_RETURN_IF_SUCCEED,
+} ebpf_return_type_t;
+```
+
+#### `ebpf_helper_function_addresses_t` Struct
+This structure is used to specify the address at which the various helper function implemented by the extension resides. If an eBPF program is JIT compiled, then the generated machine code will have `call` instructions to these addresses. For interpreted mode, the eBPF Execution Engine will invoke the functions at these addresses. The fields of this struct should be set as follows:
```

Comment:
```suggestion
This structure is used to specify the address at which the various helper functions implemented by the extension reside. If an eBPF program is JIT compiled, then the generated machine code will have `call` instructions to these addresses. For interpreted mode, the eBPF Execution Engine will invoke the functions at these addresses. The fields of this struct should be set as follows:
```
Please also split long lines into shorter lines to make github diffs easier.  Same applies to a number of lines below.

Also, what's the difference between the "Execution Engine" mentioned here, and the "Execution Context" in line 147?